python/tank/descriptor/io_descriptor/base.py

```diff
@@ -211,93 +212,94 @@ def _find_latest_tag_by_pattern(self, version_numbers, pattern):
+        # version contains all tokens until we find a "x".
+        # Separators are kept by the split being done above, so actual tokens are
+        # every other entry and we need to only consider even entries, hence the '2'
+        # step
+        for i in range(0, num_tokens, 2):
+            token = version_split[i]
+            if token == "x":
+                # If we found a "x", check that all following tokens are "x" as
+                # well, having something like v4.x.2 is illegal
+                # Tokens are every other entry, so next token is at current index+2
+                # and we only consider even entries from it, hence the '2' step.
+                for token2 in version_split[i+2:num_tokens:2]:
+                    if token2 != "x":
+                        raise TankDescriptorError(
+                            "Incorrect version pattern '%s'. "
+                            "There should be no digit after a 'x', found %s." % (
+                                pattern, token2,
+                            )
+                        )
+                # Build our base version string. Please note we will have a "." or a "-"
+                # at the end.
+                # Please note that we don't have to worry about leading "v" here: if
+                # the pattern starts with "vx" the base version will be an empty
+                # string
+                base_version = "".join(version_split[:i])
+                break
+        else:
+            # No "x" in our pattern
+            base_version = "".join(version_split)
+
+        if not base_version:
+            # Match anything, this happens if the pattern was vx.x.x
+            # Sort in reverse order, for the "v" case, hoping a valid "v" version
+            # will be in the latest ones
+            version_numbers.sort(key=LooseVersion, reverse=True)
+            if has_v_prefix:
+                # Only match versions starting with "v"
+                for version_number in version_numbers:
+                    if version_number.startswith("v"):
+                        return version_number
+                # No match
+                log.debug("Didn't find a matching version for pattern %s" % pattern)
                 return None
-            current = current[version_digit]
-            if version_to_use is None:
-                version_to_use = "v%d" % version_digit
             else:
-                version_to_use = version_to_use + ".%d" % version_digit
-
-        # at this point we have a matching version (eg. v4.x.x => v4.0.2) but
-        # there may be forked versions under this 4.0.2, so continue to recurse into
-        # the versions dictionary to find the latest forked version
-        while len(current):
-            version_digit = max(current.keys())
-            current = current[version_digit]
-            version_to_use = version_to_use + ".%d" % version_digit
-
-        return version_to_use
+                return version_numbers[0]
```

Comment:
Given that we only accept patterns with a leading "v", I went ahead and removed the `has_v_prefix` variable: patterns always have a "v" prefix.